### PR TITLE
feat(dbapi): support Connection context management usage

### PR DESCRIPTION
**Pull request by majorgreys**
## Description
The dbapi integration's patched connection did not return the traced connection when used as context manager, resulting in db calls not being traced. This PR fixes this and adds some safe guards for a variety of scenarios we see for how the `__enter__` is implemented in db libraries.

## Checklist
- [x] Entry added to release notes using [reno](https://docs.openstack.org/reno/latest/user/usage.html)
- [x] Tests provided; and/or
- [ ] Description of manual testing performed and explanation is included in the code and/or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/contrib/dbapi/__init__.py [hunk 1]
```diff
@@ -178,6 +178,44 @@ def __init__(self, conn, pin=None, cfg=None, cursor_cls=None):
         self._self_cursor_cls = cursor_cls
         self._self_config = cfg or config.dbapi2
 
+    def __enter__(self):
+        """Context management is not defined by the dbapi spec.
+
+        This means unforunately that the database clients each define their own
```
Comment (jd) [inline ddtrace/contrib/dbapi/__init__.py:184]:
```suggestion
        This means unfortunately that the database clients each define their own
```

**Review by jd — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

File: docs/spelling_wordlist.txt [hunk 2]
```diff
@@ -23,6 +23,7 @@ coroutines
 datadog
 datadoghq
 datastore
+dbapi
```
Comment (brettlangdon) [inline docs/spelling_wordlist.txt:26]: this is going to be a pain to maintain this list, is it worth it?

Comment (majorgreys) [reply docs/spelling_wordlist.txt:26]: If we want to automate spell checking, we'll need some form of a dictionary for our case. We could disable fail on warning but that would allow the scenario we had before where there were spelling errors in the public docs we publish.

Comment (brettlangdon) [reply docs/spelling_wordlist.txt:26]: are there any public technology word lists we can pull in as a starting place?

I feel like this is going to get cumbersome anytime anyone wants to introduce a change to the docs.

File: tests/contrib/mysqldb/test_mysqldb.py [hunk 3]
```diff
@@ -25,7 +26,7 @@ def tearDown(self):
         if self.conn:
             try:
                 self.conn.ping()
-            except MySQLdb.InterfaceError:
+            except (MySQLdb.InterfaceError, MySQLdb.OperationalError):
```
Comment (brettlangdon) [inline tests/contrib/mysqldb/test_mysqldb.py:29]: why this change here?

Comment (Kyle-Verhoog) [reply tests/contrib/mysqldb/test_mysqldb.py:29]: hmm not sure. Maybe @majorgreys was getting this locally? Seems like a pretty harmless change. I don't even know what this code is doing or why it's needed in the first place.

File: ddtrace/contrib/dbapi/__init__.py [hunk 4]
```diff
@@ -178,6 +178,46 @@ def __init__(self, conn, pin=None, cfg=None, cursor_cls=None):
...
 
+    def __enter__(self):
+        """Context management is not defined by the dbapi spec.
+
+        This means unfortunately that the database clients each define their own
+        implementations.
+
+        The ones we know about are:
+
+        - mysqlclient<2.0 which returns a cursor instance. >=2.0 returns a
```
Comment (brettlangdon) [inline ddtrace/contrib/dbapi/__init__.py:189]: We have others to add here no?

File: tests/contrib/mysqldb/test_mysqldb.py (hunk 3 above)
Comment (brettlangdon) [reply tests/contrib/mysqldb/test_mysqldb.py:29]: > Seems like a pretty harmless change

Yeah, just curious if it needs to be in this PR or on it's own

**Review by brettlangdon — APPROVED** [review]

File: ddtrace/contrib/dbapi/__init__.py (hunk 4 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/dbapi/__init__.py:189]: Ah yes we have since looked into others, I'll update

**Review by Kyle-Verhoog — APPROVED** [review]
